### PR TITLE
Add mic audio streaming

### PR DESCRIPTION
**Pull request by dmohanty-r7**
This PR adds a new Meterpreter stdapi extension for Mic. It adds the ability to open mic capture on a remote host and stream the audio back to Metasploit via a Meterpreter channel. The mic_start command writes the channel data to a file in real time such that it can be listened to/processed in realtime.

To review this PR you must also have the 'audio-channel' branch of mettle. Currently this works for mettle on OSX.

https://github.com/rapid7/metasploit-framework/pull/8639

### CODE REVIEW

File: mettle/src/stdapi/audio/apple_mic.m [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+#include "log.h"
+#include "ringbuf.h"
+
+#define TLV_TYPE_AUDIO_DURATION        (TLV_META_TYPE_UINT   | TLV_EXTENSIONS + 10)
+#define TLV_TYPE_AUDIO_DATA            (TLV_META_TYPE_RAW    | TLV_EXTENSIONS + 11)
+#define TLV_TYPE_AUDIO_INTERFACE_ID    (TLV_META_TYPE_UINT   | TLV_EXTENSIONS + 12)
+#define TLV_TYPE_AUDIO_INTERFACE_NAME  (TLV_META_TYPE_STRING | TLV_EXTENSIONS + 13)
+
+#define AUDIO_MAX_BUFFER_SIZE         65536
+#define AUDIO_TARGET_SAMPLE_RATE_HZ   11025
```
Comment (timwr) [inline mettle/src/stdapi/audio/apple_mic.m]: Not 44khz?

Comment (timwr) [inline mettle/src/stdapi/audio/apple_mic.m]: Oh it's a minimum, ignore me :)

File: mettle/src/stdapi/audio/apple_mic.m [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    @autoreleasepool {
+        [capture stop];
+    }
+    return tlv_packet_response_result(ctx, TLV_RESULT_SUCCESS);
+}
+
+struct tlv_packet *audio_mic_list(struct tlv_handler_ctx *ctx)
+{
+    struct tlv_packet *p = tlv_packet_response_result(ctx, TLV_RESULT_SUCCESS);
+    NSArray *devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeAudio];
```
Comment (timwr) [inline mettle/src/stdapi/audio/apple_mic.m]: I think this (and webcam_list) need an autoreleasepool too

Comment (pbarry-r7) [inline mettle/src/stdapi/audio/apple_mic.m]: Ooo, good point.  :+1:

File: mettle/src/stdapi/audio/apple_mic.m [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+#define AUDIO_TARGET_NUM_CHANNELS     1
+
+@interface AudioCapture : NSObject <AVCaptureAudioDataOutputSampleBufferDelegate>
+- (void) captureOutput: (AVCaptureOutput*) output
+ didOutputSampleBuffer: (CMSampleBufferRef) buffer
+        fromConnection: (AVCaptureConnection*) connection;
+@end
+@interface AudioCapture ()
+{
+    AVCaptureSession* session;
```
Comment (timwr) [inline mettle/src/stdapi/audio/apple_mic.m]: Annoyingly, for some reason, you can't do this on i386:
```
make TARGET=i386-apple-darwin
Building mettle for i386-apple-darwin
~/dev/git/mettle/mettle/src/stdapi/audio/apple_mic.m:25:23: error: instance variables may not be placed in class extension
    AVCaptureSession* session;
                      ^
```
I had to workaround it with this change:
https://github.com/rapid7/mettle/commit/d1f2a979044abd5ff72234e8483a27178e9dbc55#diff-c6d902dfa046d5a9253967394079fe86L24
My objective-C is a little sloppy so there might be a better fix.

File: mettle/configure.ac [hunk 4]
```diff
@@ -28,6 +28,7 @@ case $host_os in
 			[HOST_OS=osx
 			 AC_DEFINE(HAVE_DESKTOP_SCREENSHOT)
 			 AC_DEFINE(HAVE_CLIPBOARD)
+             AC_DEFINE(HAVE_MIC)
```
Comment (timwr) [inline mettle/configure.ac]: nitpick: indent

File: mettle/src/stdapi/stdapi.c [hunk 5]
```diff
@@ -33,4 +34,5 @@ void tlv_register_stdapi(struct mettle *m)
 	webcam_register_handlers(m);
 	ui_register_handlers(m);
 	clipboard_register_handlers(m);
+    audio_mic_register_handlers(m);
```
Comment (timwr) [inline mettle/src/stdapi/stdapi.c]: indent

**Comment by pbarry-r7** [comment]
Cool, thanks, @timwr!  I verified my changes no longer error on i386 and appear to still work fine (and I, too, am not sure of a snazzier way to do it in Objective-C...  :)  ).

**Comment by timwr** [comment]
Great stuff @pbarry-r7 and @dmohanty-r7 !!!
Landed this, we can hopefully iterate on a few things in the future. e.g:
- Compression or a variable bitrate (more useful for video)
- Video (with sounds :D)
- More platforms?
- Improved async behaviour

**Comment by pbarry-r7** [comment]
Well said, @timwr...!

**Review by titoel — APPROVED** [review]

**Pull request merged**